src/linemod_detect.cpp

```diff
@@ -162,24 +319,149 @@ struct Detector: public object_recognition_core::db::bases::ModelReaderBase {
+      cv::Mat_<cv::Vec3f> depth_real_model = depth_real_model_raw(rect_model);
+
+      //initialize the translation based on reference data
+      cv::Vec3f T_crop = depth_real_ref(depth_real_ref.rows / 2.0f, depth_real_ref.cols / 2.0f);
+      //add the object's depth
+      T_crop(2) += D_match;
+
+      if (!cv::checkRange(T_crop))
+        continue;
+      cv::Vec3f T_real_icp(T_crop);
+
+      //initialize the rotation based on model data
+      if (!cv::checkRange(R_match))
+        continue;
+      cv::Matx33f R_real_icp(R_match);
+
+      //get the point clouds (for both reference and model)
+      std::vector<cv::Vec3f> pts_real_model_temp;
+      std::vector<cv::Vec3f> pts_real_ref_temp;
+      float px_ratio_missing = matToVec(depth_real_model, depth_real_ref, pts_real_model_temp, pts_real_ref_temp);
+      if (px_ratio_missing > *px_match_min_)
+        continue;
+
+      //perform the first approximate ICP
+      float px_ratio_match_inliers = 0.0f;
+      float icp_dist = icpCloudToCloud(pts_real_ref_temp, pts_real_model_temp, R_real_icp, T_real_icp, px_ratio_match_inliers, 1);
+      //reject the match if the icp disntance if too big
+      if (icp_dist > *icp_dist_min_)
+        continue;
+
+      //perform a finer ICP
+      icp_dist = icpCloudToCloud(pts_real_ref_temp, pts_real_model_temp, R_real_icp, T_real_icp, px_ratio_match_inliers, 2);
+
+      //keep the object match
+      objs_.push_back(object_recognition_core::db::ObjData(pts_real_ref_temp, pts_real_model_temp, match.class_id, match.similarity, icp_dist, px_ratio_match_inliers, R_real_icp, T_crop));
+      ++iter;
+    }
+
+    //local non-maxima supression to find the best match at each position
+    int count_pass = 0;
+    std::vector <object_recognition_core::db::ObjData>::iterator it_o = objs_.begin();
+    for (; it_o != objs_.end(); it_o++)
+      if (!it_o->check_done)
+      {
+        //initialize the object to publish
+        object_recognition_core::db::ObjData *o_match = &(*it_o);
+        int size_th = static_cast<int>((float)o_match->pts_model.size()*0.85);
+        //find the best object match among near objects
+        std::vector <object_recognition_core::db::ObjData>::iterator it_o2 = it_o;
+        it_o2++;
+        for (; it_o2 != objs_.end(); it_o2++)
+          if (!it_o2->check_done)
+            if (cv::norm(o_match->t, it_o2->t) < *th_obj_dist_)
+            {
+              it_o2->check_done = true;
+              if ((it_o2->pts_model.size() > size_th) && (it_o2->icp_dist < o_match->icp_dist))
+                o_match = &(*it_o2);
+            }
+
+        //perform the final precise icp
+        float icp_px_match = 0.0f;
+        float icp_dist = icpCloudToCloud(o_match->pts_ref, o_match->pts_model, o_match->r, o_match->t, icp_px_match, 0);
+
+        if (*verbose_)
+          std::cout << o_match->match_class << " " << o_match->match_sim << " icp " << icp_dist << ", ";
+
+        //additional depth check
+        icp_dist = 0.007f;
```

Comment:
the same as sensor data, since it is the distance between points.
this distance is used to compute the ratio of inliers (points laying within a certain distance between two point clouds; In the paper, this distance is equal to the mean distance*3 and at the beginning, they use all points)